openjdk/src/test/java/org/conscrypt/RenegotiationTest.java

```diff
@@ -0,0 +1,447 @@
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import javax.net.ssl.SSLContext;
+import javax.net.ssl.SSLEngine;
+import javax.net.ssl.SSLEngineResult;
+import javax.net.ssl.SSLEngineResult.Status;
+import javax.net.ssl.SSLSocket;
+import javax.net.ssl.SSLSocketFactory;
+import libcore.java.security.TestKeyStore;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+import org.junit.runners.Parameterized.Parameter;
+import org.junit.runners.Parameterized.Parameters;
+
+/**
+ * This tests that server-initiated cipher renegotiation works properly with a Conscrypt client.
+ * BoringSSL does not support user-initiated renegotiation, so we use the JDK implementation for
+ * the server.
+ */
+@RunWith(Parameterized.class)
+public class RenegotiationTest {
+    private static final ByteBuffer EMPTY_BUFFER = ByteBuffer.allocateDirect(0);
+    private static final String[] CIPHERS = TestUtils.getCommonCipherSuites();
+    private static final byte[] MESSAGE_BYTES = "Hello".getBytes(TestUtils.UTF_8);
+    private static final ByteBuffer MESSAGE_BUFFER =
+            ByteBuffer.wrap(MESSAGE_BYTES).asReadOnlyBuffer();
+    private static final int MESSAGE_LENGTH = MESSAGE_BYTES.length;
+    private static final int NUM_REPLIES = 1;
+
+    public enum SocketType {
+        FILE_DESCRIPTOR {
+            @Override
+            Client newClient(int port) {
+                return new Client(false, port);
+            }
+        },
+        ENGINE {
+            @Override
+            Client newClient(int port) {
+                return new Client(true, port);
+            }
+        };
+
+        abstract Client newClient(int port);
+    }
+
+    @Parameters(name = "{0}")
+    public static Object[] data() {
+        return new Object[] {SocketType.FILE_DESCRIPTOR, SocketType.ENGINE};
+    }
+
+    @Parameter
+    public SocketType socketType;
+
+    private Client client;
+    private Server server;
+
+    @Before
+    public void setup() throws Exception {
+        server = new Server();
+        Future<?> connectedFuture = server.start();
+
+        client = socketType.newClient(server.port());
+        client.start();
+
+        // Wait for the initial connection to complete.
+        connectedFuture.get(5, TimeUnit.SECONDS);
+    }
+
+    @After
+    public void teardown() {
+        client.stop();
+        server.stop();
+    }
+
+    @Test
+    public void test() throws Exception {
+        client.sendMessage();
+
+        Future<?> repliesFuture = client.readReplies();
+        server.await(5, TimeUnit.SECONDS);
+        repliesFuture.get(5, TimeUnit.SECONDS);
+
+        // BoringSSL will not call the verify callback on renegotiation, so the client will never
```

Comment:
This seems like a problem.  Won't our getters just be lying then?  I don't know if there's any BoringSSL changes coming on this or anything, but having getCipherSuite() (and maybe others) returning values that are just untrue seems like a bad idea.

Reply:
@flooey this was actually due to the fact that we were caching the cipher in our `ActiveSession`. That seems to be the only potential caching issue, since BoringSSL won't let the other values change (peer cipher suite, etc.). This should b fixed now.